### PR TITLE
Added nestjs module for notifire

### PR DESCRIPTION
**Pull request by devblin**
- **What kind of change does this PR introduce?**
Implements feature #74

### CODE REVIEW

**Comment by scopsy** [comment]
@devblin This looks great, I'll take a closer look I believe tomorrow and play with it a bit. Thank you so much for this! Super exciting.

File: packages/nest/package.json [hunk 1]
```diff
@@ -0,0 +1,89 @@
+{
+  "name": "@notifire/notifire",
+  "version": "0.2.3",
+  "description": "A notifire wrapper for notifire",
```
Comment (scopsy) [inline packages/nest/package.json:4]:
```suggestion
  "description": "A nestjs wrapper for notifire",
```

File: packages/nest/src/interfaces/notifire-options.interface.ts [hunk 2]
```diff
@@ -0,0 +1,9 @@
+// eslint-disable-next-line @typescript-eslint/no-empty-interface
+export interface NotifireOptions {
+  //
+  // This interface describes the options you want to pass to
+  // NotifireModule.
+  //
+  providers: Array<any>;
```
Comment (scopsy) [inline packages/nest/src/interfaces/notifire-options.interface.ts:7]: Can we import the IProvider and ITemplate interface from @notifire/core here ?

File: packages/nest/src/services/notifire.service.ts [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+    if (!this.notifire) {
+      this.notifire = new Notifire();
+    }
+    return this.notifire;
+  }
+
+  async registerProviders(): Promise<any> {
+    const providerPromises = [];
+
+    for (let i = 0; i < this.providers.length; i += 1) {
```
Comment (scopsy) [inline packages/nest/src/services/notifire.service.ts:30]: Can be simplified to:
```suggestion
    for (const provider of this.providers.length) {
```

File: packages/nest/src/services/notifire.service.ts [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+      const waiting = await this.notifire.registerProvider(this.providers[i]);
```
Comment (scopsy) [inline packages/nest/src/services/notifire.service.ts:31]: M
```suggestion
      await this.notifire.registerProvider(provider);
```

File: packages/nest/src/services/notifire.service.ts [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+      providerPromises.push(waiting);
+    }
+
+    return await Promise.all(providerPromises);
```
Comment (scopsy) [inline packages/nest/src/services/notifire.service.ts:35]: Not sure you need the Promise.all here if we awaited in the for loop already.

File: packages/nest/src/services/notifire.service.ts [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+  }
+
+  async registerTemplates(): Promise<any> {
+    const templatePromises = [];
+
+    for (let i = 0; i < this.templates.length; i += 1) {
+      const waiting = await this.notifire.registerTemplate(this.templates[i]);
```
Comment (scopsy) [inline packages/nest/src/services/notifire.service.ts:42]: Same goes here aswell

**Review by scopsy — COMMENTED** [review]
Looks very good @devblin ! Could you please send me an invite to your fork? I have some small changes wanted to commit but seems like I Can't push them

**Comment by devblin** [comment]
> Looks very good @devblin ! Could you please send me an invite to your fork? I have some small changes wanted to commit but seems like I Can't push them

Yeah sure

**Comment by devblin** [comment]
@scopsy I have sent you an invite.

File: packages/nest/src/interfaces/notifire-options.interface.ts (hunk 2 above)
Comment (devblin) [reply packages/nest/src/interfaces/notifire-options.interface.ts:7]: Yes, we can.

**Comment by devblin** [comment]
@scopsy Can I start commiting the changes you mentioned ?

**Comment by scopsy** [comment]
@devblin sure!

**Comment by devblin** [comment]
@scopsy I have commited the changes and tested locally.

**Comment by scopsy** [comment]
Amazing, I'll try to play with it over the weekend. Nice work!!

File: packages/nest/README.md [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+
+Using notifire's singleton service in other services and modules:
+
+```javascript
+import { Injectable, Inject } from '@nestjs/common';
+import { NotifireService } from '@notifire/nest';
+
+@Injectable()
+export class UserService {
+  constructor(private readonly notifire: NotifireService) {}
```
Comment (scopsy) [inline packages/nest/README.md:49]: @devblin I've simplified the improt so we don't have use `@Inject` what do you think ?

**Comment by scopsy** [comment]
@devblin I've simplified the injection system a bit, and tested your pr against our nodejs API, looks awesome! 

I think we can merge this to `v0.3.0` once you give me the green lights :)

**Comment by devblin** [comment]
> @devblin I've simplified the injection system a bit, and tested your pr against our nodejs API, looks awesome!
> 
> I think we can merge this to `v0.3.0` once you give me the green lights :)

We are good to go 🥳

**Comment by davidsoderberg** [comment]
Looks good to me 🥳

File: packages/nest/README.md [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+    @Module({
+      imports: [
+        NotifireModule.forRoot({
+          providers: [
+            new SendgridEmailProvider({
+              apiKey: process.env.SENDGRID_API_KEY,
+              from: 'sender@mail.com',
+            }),
+          ],
+          templates: [
```
Comment (evgenyidf) [inline packages/nest/README.md:20]: Is that possible to declare providers' & templates separately ? 
for example service constructor ?

Comment (scopsy) [reply packages/nest/README.md:20]: @evgenyidf yeah sure, you can just use the NotifireService and call `regiserProvider` or `regiserTemplate` as usual.

**Pull request merged**